DHIS2: Add support for sending geolocation

Charl1996 opened the pull request:
## Summary
[Ticket](https://dimagi-dev.atlassian.net/browse/SC-691)

This PR adds support for sending geolocation information on `events` and `enrolments` of TrackedEntities.
[Doc](https://docs.dhis2.org/en/develop/using-the-api/dhis-core-version-236/tracker.html#webapi_tracker_api)

## Feature Flag
DHIS2

## Product description
A [config setting](https://staging.commcarehq.org/a/ben-test/motech/forwarding/config/Dhis2EntityRepeater/cd3f680f288829c5a7545bfc3ce61e65/) (`event_location`) is added to DHIS2 `form_configs` when configuring DHIS2 Tracked Entities and / or Anonymous Events. The app builders should decide where they want the event location to come from (i.e. `form_question`, `case_property` etc.).   

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage
Unit tests included for parsing location data when forwarding cases to DHIS2

### Safety story
Unit tests included

Tests performed on Staging:
- Location configured and captured from form_question: [here's](https://staging.commcarehq.org/a/ben-test/motech/logs/520019/) the Remote API response as the location is being sent to DHIS2 for event and enrollment.
- No location or configuration specified (tested for backwards compatibility) - response [here](https://staging.commcarehq.org/a/ben-test/motech/logs/520013/) (same as would have been previously). Thus, this change should not affect current projects.

 

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

Charl1996 committed:
Add coordinate to DHISFormConfig

Charl1996 committed:
Add geometry to enrollment

Charl1996 committed:
Test location parsing

Charl1996 committed:
Add support for feature type retrieval

Charl1996 committed:
Remove redundant comment

Charl1996 commented on corehq/motech/dhis2/dhis2_config.py:
```diff
@@ -46,6 +46,9 @@ class Dhis2FormConfig(DocumentSchema):
     completed_date = DictProperty(required=False)
     datavalue_maps = SchemaListProperty(FormDataValueMap)
+    coordinate = DictProperty(required=False, default={
+        "form_question": "/metadata/location/#text"
```
I determined this default value from actually inspecting a case which had the location being sent in the background.

Charl1996 commented:
I'm still trying to figure out the best way to send any past cases' locations through...

kaapstorm commented:
> the best way to send any past cases' locations through

What about changing something on the case, to trigger a payload? For an individual case, this could be done by using the "Clean Case Data" button in HQ, or for many cases by using Case Import. I think a user/admin could even just set a value back to itself -- wouldn't need to change the case at all.

Would that work?

Charl1996 commented:
@kaapstorm I was thinking of changing the repeat record's state to let it trigger, but then I'd have to write a migration...changing somthing on the case on the other hand requires manual labour for every domain?

(I'm awaiting a response on the ticket if it's important to consider past cases' location data to be sent to DHIS2 - can't see why not, really, but if it is not important for some reason, I won't bother)

kaapstorm commented on corehq/motech/dhis2/events_helpers.py:
```diff
@@ -32,6 +32,7 @@ def get_event(domain, config, form_json=None, info=None):
         _get_event_status,
         _get_completed_date,
         _get_datavalues,
+        _get_coordinates
```
Ultranit: Missing comma.

kaapstorm reviewed (commented):
What happens to the payloads of existing Anonymous Events and Tracked Entity integrations when this change is rolled out? In other words, how will this affect existing projects?

I noticed that this is deployed to Staging. If you include what manual testing you've done under "Safety Story", it will improve confidence in this change.

kaapstorm commented:
> changing the repeat record's state to let it trigger

I think it's totally OK to say that this feature is available from now on, without trying to change historical data.

If projects do want us to change historical data, there might be bigger-picture implications. What is the DHIS2 data used for? How does changing old data affect those uses? Always good? Sometimes good, sometimes not? A thing that one group thinks is a great idea but has a detrimental downstream effect?

We can resend repeat records using the repeat records report. (Remember that repeat records are identified by a form submission ID, not a case ID.)

kaapstorm commented:
> ... changing something on the case through the portal on the other hand requires manual labour for every domain?

Resending every Event and every Tracked Entity ever sent to DHIS2 for every domain is something I would definitely not do. If there is a project that wants to resend historical data, and we know that it is safe (in a bigger-picture, human, non-IT sense) to do that, then I'd suggest writing a management command for changing those repeat records to "pending".

Charl1996 commented on corehq/motech/dhis2/entities_helpers.py:
```diff
@@ -392,7 +392,7 @@ def get_tracked_entity_type(requests, entity_type_id):
 
 def get_geo_json(form_config, case_trigger_info, entity_feature_type=None):
-    feature_type = 'NONE'  # feature_type = 'NONE' for most entities, including 'Person'
+    feature_type = 'NONE'
```
Yes, it's a DHIS2 thing. [This doc](https://docs.dhis2.org/en/develop/using-the-api/dhis-core-version-236/tracker.html#webapi_tracker_api) (see `Creating a new tracked entity instance`) specifies that the featureType of the GeoJson must match the featureType of the entityType. I had to use Postman to get all the featureTypes. `'None'` seems to be a very popular featureType, so I figured I'd default to that.

(come to think of it, maybe I'll extract that to a constant with a descriptive name)

Charl1996 commented on corehq/motech/dhis2/events_helpers.py:
```diff
@@ -110,30 +110,23 @@ def _get_coordinates(config, case_trigger_info):
+    coordinate: "-35.8655497 14.6941185 138.66 5.4"
+    Returns: {"latitude": "-35.8655", "longitude": "14.6941"} conforming to EPSG:4326
+
+    Notes
+    According to the documentation no more than a maximum of 4
+    significant decimal places should ever be necessary:
+    https://docs.dhis2.org/en/use/user-guides/dhis-core-version-234/configuring-the-system/maps.html#gis_creating_setup
```
Thanks for the suggestion - I'll maybe spend some time to read up on RST. :+1:

Charl1996 commented on corehq/motech/dhis2/events_helpers.py:
```diff
@@ -99,6 +100,29 @@ def _get_datavalues(config, case_trigger_info):
+    """
+    Example
+    coordinate: "-35.8655497 14.6940185 138.66 5.4"
+    Returns: {"latitude": -35.8655497, "longitude": 14.6940185}
```
@kaapstorm Absolutely! For some reason I never think of using doctests!

Charl1996 commented:
@kaapstorm 

> Does "/metadata/location/#text" always exist on a form submission?

No, but it seems to be the default when an app is configured to send the location for forms in the background. When the mobile user has to actually submit a question to capture the location, this value can and should be changed for the [Case Config](https://staging.commcarehq.org/a/ben-test/motech/forwarding/config/Dhis2EntityRepeater/a91580d0c8e5c9615eeae941a78a281a/). 

I did not consider the case where the app is configured to always send location in background (in this case the location data sits on app metadata). I'm guessing we could support this, but I thought it reasonable that if the patient location is important, the app needs to be configured to send in background on each submission or as a question. Any other thoughts?

> Does this change affect the UI?

In a small way. The [`form_configs`](https://staging.commcarehq.org/a/ben-test/motech/forwarding/config/Dhis2EntityRepeater/a91580d0c8e5c9615eeae941a78a281a/) now has the default `coordinate` setting.

(I need to update the PR description - don't know why I didn't include this)

> Can users set this to something else? 

Yes

> Should they be allowed to? 

Yes

> Is this a constant?

No, can/should be changed in case of location captured by question.

Charl1996 commented:
> What happens to the payloads of existing Anonymous Events and Tracked Entity integrations when this change is rolled out? In other words, how will this affect existing projects?

Do you mean those that are being retried?

Charl1996 commented:
> If you include what manual testing you've done under "Safety Story", it will improve confidence in this change.

I have as of yet only confirmed that the location data is being retrieved correctly (by executing certain methods myself on staging with the configured data and submitted form), but still busy with trying for a more general e2e test.

Charl1996 committed:
Use mobile user location as events/enrolments location reference

Charl1996 committed:
Fix str -> float issue

Charl1996 committed:
Fix geoJson type

Charl1996 committed:
Remove redundant constant

Charl1996 committed:
Fix accessor method

Charl1996 commented:
I've updated this PR quite significantly. Turns out my initial understanding of where the location information should come from was misaligned. 

Essential changes:
- Grab location from actual location the user belongs to and not the mobile submitted location from the form.
- Changed the GeoJson type for enrollments to hardcode "Point" - according to [this](https://docs.dhis2.org/en/develop/using-the-api/dhis-core-version-236/tracker.html#webapi_tracker_api) doc:
> the GeoJson have to match the featureType of the TrackedEntityType definition

I initially interpreted it as the literal type, i.e. EntityType "Person" has featureType "NONE", but DHIS2 does not accept that type for the GeoJson...which...in retrospect makes sense...a GeoJson specifies geographical features, so for a single point like this would be, the type "Point" is appropriate. Other types include "Polygon" and "LineString", but for this use-case only "Point" would ever be appropriate.

Charl1996 committed:
Move location extraction to form config

Charl1996 committed:
Update README

kaapstorm reviewed (approved):
This has turned out really nicely.

snopoke commented on corehq/motech/dhis2/tests/test_entities_helpers.py:
```diff
@@ -43,6 +59,114 @@ def test_bad_data_type(self):
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.user.delete(deleted_by=None)
```
This is missing the `deleted_by_domain` argument (https://github.com/dimagi/commcare-hq/blob/cc0484b7eb84ece50b8a323c4be1e0e95bb8eccf/corehq/apps/users/models.py#L1098). I'm not quite sure how the build passed.

snopoke replied:
https://github.com/dimagi/commcare-hq/pull/30314

Charl1996 replied:
Woah, this is strange. This had to pass local testing as well as Travis...interesting...
